File: templates/problems/05_binary_search/06_time_based_kv/python/solution.py
import bisect


class SolveTimeMap:
    """Time-based key-value store.

    Set: O(1), Get: O(log n). Space: O(n).
    """

    def __init__(self) -> None:
        self._store: dict[str, list[tuple[int, str]]] = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        self._store.setdefault(key, []).append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        entries = self._store.get(key, [])
        if not entries:
            return ""
        # Binary search for the largest timestamp <= given timestamp.
        # bisect_right gives the index after the last timestamp <= target.
        idx = bisect.bisect_right(entries, (timestamp, chr(0x10FFFF)))
        if idx == 0:
            return ""
        return entries[idx - 1][1]

File: templates/problems/05_binary_search/06_time_based_kv/go/solution.go
package time_based_kv

import "sort"

// SolveTimeMap implements a time-based key-value store. Set is O(1), Get is O(log n). O(n) space.
type SolveTimeMap struct {
	store map[string][]entry
}

type entry struct {
	timestamp int
	value     string
}

func NewSolveTimeMap() SolveTimeMap {
	return SolveTimeMap{store: make(map[string][]entry)}
}

func (t *SolveTimeMap) Set(key string, value string, timestamp int) {
	t.store[key] = append(t.store[key], entry{timestamp, value})
}

func (t *SolveTimeMap) Get(key string, timestamp int) string {
	entries := t.store[key]
	if len(entries) == 0 {
		return ""
	}
	// Binary search for the largest timestamp <= given timestamp
	idx := sort.Search(len(entries), func(i int) bool {
		return entries[i].timestamp > timestamp
	})
	if idx == 0 {
		return ""
	}
	return entries[idx-1].value
}
